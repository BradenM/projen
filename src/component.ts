import { Construct } from "constructs";

/**
 * Represents a project component.
 */
export class Component extends Construct {
  constructor(scope: Construct, id: string) {
    super(scope, id);
  }

  /**
   * Called before synthesis.
   */
  public preSynthesize() {}

  /**
   * Synthesizes files to the project output directory.
   */
  public synthesize() {}

  /**
   * Called after synthesis. Order is *not* guaranteed.
   */
  public postSynthesize() {}
}
