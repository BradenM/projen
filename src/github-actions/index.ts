export * from "./actions";
export * from "./github-action-metadata";
