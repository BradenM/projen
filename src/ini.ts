import { Construct } from "constructs";
import * as INI from "ini";
import { IResolver } from "./file";
import { ObjectFile, ObjectFileOptions } from "./object-file";

/**
 * Options for `IniFile`.
 */
export interface IniFileOptions extends ObjectFileOptions {}

/**
 * Represents an INI file.
 */
export class IniFile extends ObjectFile {
  constructor(scope: Construct, filePath: string, options: IniFileOptions) {
    super(scope, filePath, options);
  }

  protected synthesizeContent(resolver: IResolver): string | undefined {
    const json = super.synthesizeContent(resolver);
    if (!json) {
      return undefined;
    }

    return [
      ...(this.marker ? [`# ${this.marker}`] : []),
      "",
      INI.stringify(JSON.parse(json)),
    ].join("\n");
  }
}
