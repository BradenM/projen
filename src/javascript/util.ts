import { readFileSync } from "fs";
import { basename, dirname, extname, join, sep } from "path";
import * as semver from "semver";

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>\w+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifestId = `${dependencyName}/package.json`;
  try {
    const manifestPath = require.resolve(manifestId, options);
    // cannot just `require('dependency/package.json')` here because
    // `options.paths` may not overlap with this node proc's resolution paths.
    const manifest = JSON.parse(
      readFileSync(manifestPath, "utf-8").toString()
    ) as {
      version: string;
    };
    return manifest.version;
  } catch (e) {
    return undefined;
  }
}
