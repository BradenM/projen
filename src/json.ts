import * as path from "path";
import { Construct, IConstruct } from "constructs";
import { IResolver } from "./file";
import { ObjectFile, ObjectFileOptions } from "./object-file";
import { Project } from "./project";

/**
 * Options for `JsonFile`.
 */
export interface JsonFileOptions extends ObjectFileOptions {
  /**
   * Adds a newline at the end of the file.
   * @default true
   */
  readonly newline?: boolean;
}

/**
 * Represents a JSON file.
 */
export class JsonFile extends ObjectFile {
  /**
   * Finds a JSON file by name in the given scope.
   * @param filePath The file path. If this path is relative, it will be resolved
   * from the root of the nearest project.
   */
  public static tryFindJsonFile(
    scope: IConstruct,
    filePath: string
  ): ObjectFile | undefined {
    const isObjectFile = (c: IConstruct): c is ObjectFile =>
      c instanceof ObjectFile;
    const absolutePath = path.isAbsolute(filePath)
      ? filePath
      : path.resolve(Project.ofProject(scope).outdir, filePath);
    return scope.node
      .findAll()
      .filter(isObjectFile)
      .find((file) => file.absolutePath === absolutePath);
  }

  private readonly newline: boolean;

  constructor(scope: Construct, filePath: string, options: JsonFileOptions) {
    super(scope, filePath, options);

    this.newline = options.newline ?? true;

    if (!options.obj) {
      throw new Error('"obj" cannot be undefined');
    }
  }

  protected synthesizeContent(resolver: IResolver): string | undefined {
    const json = super.synthesizeContent(resolver);
    if (!json) {
      return undefined;
    }

    const sanitized = JSON.parse(json);

    if (this.marker) {
      sanitized["//"] = this.marker;
    }

    let content = JSON.stringify(sanitized, undefined, 2);
    if (this.newline) {
      content += "\n";
    }

    return content;
  }
}
