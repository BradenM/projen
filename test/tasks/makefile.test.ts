import { Project } from '../../src';
import { TasksEngine } from '../../src/tasks';
import { synthSnapshot, TestProject } from '../../src/util/synth';

test('empty task', () => {
  // GIVEN
  const p = new TestProject();

  // WHEN
  p.addTask('empty');

  // THEN
  const makefile = synthMakefile(p);
  expect(makefile).toMatchSnapshot();
});

test('multiple "exec" commands', () => {
  // GIVEN
  const p = new TestProject({ tasksOptions: { engine: TasksEngine.MAKE } });

  // WHEN
  const task = p.addTask('hello', {
    description: 'hello, world',
    exec: 'echo hello', // initial command
    env: {
      FOO: 'bar',
    },
  });

  task.exec('echo world');
  task.exec('echo "with quotes"');
  task.env('BAR', 'baz');

  // THEN
  const makefile = synthMakefile(p);
  expect(makefile).toMatchSnapshot();
});

test('subtasks', () => {
  // GIVEN
  const p = new TestProject({ tasksOptions: { engine: TasksEngine.MAKE } });
  const hello = p.addTask('hello', { exec: 'echo hello' });
  const world = p.addTask('world');

  // WHEN
  world.exec('echo "running hello"');
  world.spawn(hello);

  // THEN
  const makefile = synthMakefile(p);
  expect(makefile).toMatchSnapshot();
});

test('env() can be used to add environment variables', () => {
  // GIVEN
  const p = new TestProject({ tasksOptions: { engine: TasksEngine.MAKE } });
  const t = p.addTask('my-task', {
    env: {
      INITIAL: '123',
      ENV: '456',
    },
  });

  // WHEN
  t.env('FOO', 'BAR');
  t.env('HELLO', 'world');

  // THEN
  const makefile = synthMakefile(p);
  expect(makefile).toMatchSnapshot();
});

test('"condition" can be used to define a command that will determine if a task should be skipped', () => {
  // GIVEN
  const p = new TestProject({ tasksOptions: { engine: TasksEngine.MAKE } });

  // WHEN
  p.addTask('foo', {
    condition: 'false',
    exec: 'foo bar',
  });

  // THEN
  const makefile = synthMakefile(p);
  expect(makefile).toMatchSnapshot();
});

test('"builtin" can be used to execute builtin commands', () => {
  // GIVEN
  const p = new TestProject({ tasksOptions: { engine: TasksEngine.MAKE } });
  const task = p.addTask('foo', {
    condition: 'false',
  });

  // WHEN
  task.builtin('tasks/builtin-example');

  // THEN
  const makefile = synthMakefile(p) as string;

  // sanitize output - the test runs in a random temporary directory, so the
  // resolved path to the builtin will not be consistent
  const sanitized = makefile.replace(/node .*lib\//, 'node lib/');

  expect(sanitized).toMatchSnapshot();
});

test('"requiredEnv" can be used to specify required environment variables', () => {
  // GIVEN
  const p = new TestProject({ tasksOptions: { engine: TasksEngine.MAKE } });

  // WHEN
  p.addTask('foo', {
    requiredEnv: ['MISSING1', 'MISSING2', 'NOT_MISSING'],
  });

  // THEN
  const makefile = synthMakefile(p);
  expect(makefile).toMatchSnapshot();
});

function synthMakefile(p: Project) {
  return synthSnapshot(p).Makefile;
}