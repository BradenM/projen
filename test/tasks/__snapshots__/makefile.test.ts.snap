// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"builtin" can be used to execute builtin commands 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

ROOT := $(dir $(firstword $(MAKEFILE_LIST)))

.PHONY: all
all: default

.PHONY: build
build:                        ## Full release build
	@>&2 echo 🤖 Running task \\\\\\\\033[32mbuild\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	make default ;\\\\
	cd $(ROOT) ;\\\\
	make pre-compile ;\\\\
	cd $(ROOT) ;\\\\
	make compile ;\\\\
	cd $(ROOT) ;\\\\
	make post-compile ;\\\\
	cd $(ROOT) ;\\\\
	make test ;\\\\
	cd $(ROOT) ;\\\\
	make package ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mbuild\\\\\\\\033[0m. ;\\\\

.PHONY: compile
compile:                      ## Only compile
	@>&2 echo 🤖 Running task \\\\\\\\033[32mcompile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mcompile\\\\\\\\033[0m. ;\\\\

.PHONY: default
default:                      ## Synthesize project files
	@>&2 echo 🤖 Running task \\\\\\\\033[32mdefault\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mdefault\\\\\\\\033[0m. ;\\\\

.PHONY: foo
foo:                          ## No description
	@>&2 echo 🤖 Running task \\\\\\\\033[32mfoo\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	false || { >&2 echo \\"condition exited with non-zero - exiting\\" && exit 0; } ;\\\\
	cd $(ROOT) ;\\\\
	node lib/tasks/builtin-example.task.js ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mfoo\\\\\\\\033[0m. ;\\\\

.PHONY: package
package:                      ## Creates the distribution package
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpackage\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpackage\\\\\\\\033[0m. ;\\\\

.PHONY: post-compile
post-compile:                 ## Runs after successful compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m. ;\\\\

.PHONY: pre-compile
pre-compile:                  ## Prepare the project for compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m. ;\\\\

.PHONY: test
test:                         ## Run tests
	@>&2 echo 🤖 Running task \\\\\\\\033[32mtest\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mtest\\\\\\\\033[0m. ;\\\\

.PHONY: help
help:                         ## Show help messages for make targets
	@echo \\"\\\\033[1;39mCOMMANDS:\\\\033[0m\\"; grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = \\":.*?## \\"}; {printf \\"\\\\t\\\\033[32m%-30s\\\\033[0m %s\\\\n\\", $$1, $$2}'
"
`;

exports[`"condition" can be used to define a command that will determine if a task should be skipped 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

ROOT := $(dir $(firstword $(MAKEFILE_LIST)))

.PHONY: all
all: default

.PHONY: build
build:                        ## Full release build
	@>&2 echo 🤖 Running task \\\\\\\\033[32mbuild\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	make default ;\\\\
	cd $(ROOT) ;\\\\
	make pre-compile ;\\\\
	cd $(ROOT) ;\\\\
	make compile ;\\\\
	cd $(ROOT) ;\\\\
	make post-compile ;\\\\
	cd $(ROOT) ;\\\\
	make test ;\\\\
	cd $(ROOT) ;\\\\
	make package ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mbuild\\\\\\\\033[0m. ;\\\\

.PHONY: compile
compile:                      ## Only compile
	@>&2 echo 🤖 Running task \\\\\\\\033[32mcompile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mcompile\\\\\\\\033[0m. ;\\\\

.PHONY: default
default:                      ## Synthesize project files
	@>&2 echo 🤖 Running task \\\\\\\\033[32mdefault\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mdefault\\\\\\\\033[0m. ;\\\\

.PHONY: foo
foo:                          ## No description
	@>&2 echo 🤖 Running task \\\\\\\\033[32mfoo\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	false || { >&2 echo \\"condition exited with non-zero - exiting\\" && exit 0; } ;\\\\
	cd $(ROOT) ;\\\\
	foo bar ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mfoo\\\\\\\\033[0m. ;\\\\

.PHONY: package
package:                      ## Creates the distribution package
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpackage\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpackage\\\\\\\\033[0m. ;\\\\

.PHONY: post-compile
post-compile:                 ## Runs after successful compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m. ;\\\\

.PHONY: pre-compile
pre-compile:                  ## Prepare the project for compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m. ;\\\\

.PHONY: test
test:                         ## Run tests
	@>&2 echo 🤖 Running task \\\\\\\\033[32mtest\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mtest\\\\\\\\033[0m. ;\\\\

.PHONY: help
help:                         ## Show help messages for make targets
	@echo \\"\\\\033[1;39mCOMMANDS:\\\\033[0m\\"; grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = \\":.*?## \\"}; {printf \\"\\\\t\\\\033[32m%-30s\\\\033[0m %s\\\\n\\", $$1, $$2}'
"
`;

exports[`"requiredEnv" can be used to specify required environment variables 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

ROOT := $(dir $(firstword $(MAKEFILE_LIST)))

.PHONY: all
all: default

.PHONY: build
build:                        ## Full release build
	@>&2 echo 🤖 Running task \\\\\\\\033[32mbuild\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	make default ;\\\\
	cd $(ROOT) ;\\\\
	make pre-compile ;\\\\
	cd $(ROOT) ;\\\\
	make compile ;\\\\
	cd $(ROOT) ;\\\\
	make post-compile ;\\\\
	cd $(ROOT) ;\\\\
	make test ;\\\\
	cd $(ROOT) ;\\\\
	make package ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mbuild\\\\\\\\033[0m. ;\\\\

.PHONY: compile
compile:                      ## Only compile
	@>&2 echo 🤖 Running task \\\\\\\\033[32mcompile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mcompile\\\\\\\\033[0m. ;\\\\

.PHONY: default
default:                      ## Synthesize project files
	@>&2 echo 🤖 Running task \\\\\\\\033[32mdefault\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mdefault\\\\\\\\033[0m. ;\\\\

.PHONY: foo
foo:                          ## No description
	@>&2 echo 🤖 Running task \\\\\\\\033[32mfoo\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mfoo\\\\\\\\033[0m. ;\\\\

.PHONY: package
package:                      ## Creates the distribution package
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpackage\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpackage\\\\\\\\033[0m. ;\\\\

.PHONY: post-compile
post-compile:                 ## Runs after successful compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m. ;\\\\

.PHONY: pre-compile
pre-compile:                  ## Prepare the project for compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m. ;\\\\

.PHONY: test
test:                         ## Run tests
	@>&2 echo 🤖 Running task \\\\\\\\033[32mtest\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mtest\\\\\\\\033[0m. ;\\\\

.PHONY: help
help:                         ## Show help messages for make targets
	@echo \\"\\\\033[1;39mCOMMANDS:\\\\033[0m\\"; grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = \\":.*?## \\"}; {printf \\"\\\\t\\\\033[32m%-30s\\\\033[0m %s\\\\n\\", $$1, $$2}'
"
`;

exports[`empty task 1`] = `undefined`;

exports[`env() can be used to add environment variables 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

ROOT := $(dir $(firstword $(MAKEFILE_LIST)))

.PHONY: all
all: default

.PHONY: build
build:                        ## Full release build
	@>&2 echo 🤖 Running task \\\\\\\\033[32mbuild\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	make default ;\\\\
	cd $(ROOT) ;\\\\
	make pre-compile ;\\\\
	cd $(ROOT) ;\\\\
	make compile ;\\\\
	cd $(ROOT) ;\\\\
	make post-compile ;\\\\
	cd $(ROOT) ;\\\\
	make test ;\\\\
	cd $(ROOT) ;\\\\
	make package ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mbuild\\\\\\\\033[0m. ;\\\\

.PHONY: compile
compile:                      ## Only compile
	@>&2 echo 🤖 Running task \\\\\\\\033[32mcompile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mcompile\\\\\\\\033[0m. ;\\\\

.PHONY: default
default:                      ## Synthesize project files
	@>&2 echo 🤖 Running task \\\\\\\\033[32mdefault\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mdefault\\\\\\\\033[0m. ;\\\\

.PHONY: my-task
my-task:                      ## No description
	@>&2 echo 🤖 Running task \\\\\\\\033[32mmy-task\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	INITIAL=123 ;\\\\
	ENV=456 ;\\\\
	FOO=BAR ;\\\\
	HELLO=world ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mmy-task\\\\\\\\033[0m. ;\\\\

.PHONY: package
package:                      ## Creates the distribution package
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpackage\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpackage\\\\\\\\033[0m. ;\\\\

.PHONY: post-compile
post-compile:                 ## Runs after successful compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m. ;\\\\

.PHONY: pre-compile
pre-compile:                  ## Prepare the project for compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m. ;\\\\

.PHONY: test
test:                         ## Run tests
	@>&2 echo 🤖 Running task \\\\\\\\033[32mtest\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mtest\\\\\\\\033[0m. ;\\\\

.PHONY: help
help:                         ## Show help messages for make targets
	@echo \\"\\\\033[1;39mCOMMANDS:\\\\033[0m\\"; grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = \\":.*?## \\"}; {printf \\"\\\\t\\\\033[32m%-30s\\\\033[0m %s\\\\n\\", $$1, $$2}'
"
`;

exports[`multiple "exec" commands 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

ROOT := $(dir $(firstword $(MAKEFILE_LIST)))

.PHONY: all
all: default

.PHONY: build
build:                        ## Full release build
	@>&2 echo 🤖 Running task \\\\\\\\033[32mbuild\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	make default ;\\\\
	cd $(ROOT) ;\\\\
	make pre-compile ;\\\\
	cd $(ROOT) ;\\\\
	make compile ;\\\\
	cd $(ROOT) ;\\\\
	make post-compile ;\\\\
	cd $(ROOT) ;\\\\
	make test ;\\\\
	cd $(ROOT) ;\\\\
	make package ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mbuild\\\\\\\\033[0m. ;\\\\

.PHONY: compile
compile:                      ## Only compile
	@>&2 echo 🤖 Running task \\\\\\\\033[32mcompile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mcompile\\\\\\\\033[0m. ;\\\\

.PHONY: default
default:                      ## Synthesize project files
	@>&2 echo 🤖 Running task \\\\\\\\033[32mdefault\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mdefault\\\\\\\\033[0m. ;\\\\

.PHONY: hello
hello:                        ## hello, world
	@>&2 echo 🤖 Running task \\\\\\\\033[32mhello\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	FOO=bar ;\\\\
	BAR=baz ;\\\\
	cd $(ROOT) ;\\\\
	echo hello ;\\\\
	cd $(ROOT) ;\\\\
	echo world ;\\\\
	cd $(ROOT) ;\\\\
	echo \\"with quotes\\" ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mhello\\\\\\\\033[0m. ;\\\\

.PHONY: package
package:                      ## Creates the distribution package
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpackage\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpackage\\\\\\\\033[0m. ;\\\\

.PHONY: post-compile
post-compile:                 ## Runs after successful compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m. ;\\\\

.PHONY: pre-compile
pre-compile:                  ## Prepare the project for compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m. ;\\\\

.PHONY: test
test:                         ## Run tests
	@>&2 echo 🤖 Running task \\\\\\\\033[32mtest\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mtest\\\\\\\\033[0m. ;\\\\

.PHONY: help
help:                         ## Show help messages for make targets
	@echo \\"\\\\033[1;39mCOMMANDS:\\\\033[0m\\"; grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = \\":.*?## \\"}; {printf \\"\\\\t\\\\033[32m%-30s\\\\033[0m %s\\\\n\\", $$1, $$2}'
"
`;

exports[`subtasks 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

ROOT := $(dir $(firstword $(MAKEFILE_LIST)))

.PHONY: all
all: default

.PHONY: build
build:                        ## Full release build
	@>&2 echo 🤖 Running task \\\\\\\\033[32mbuild\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	make default ;\\\\
	cd $(ROOT) ;\\\\
	make pre-compile ;\\\\
	cd $(ROOT) ;\\\\
	make compile ;\\\\
	cd $(ROOT) ;\\\\
	make post-compile ;\\\\
	cd $(ROOT) ;\\\\
	make test ;\\\\
	cd $(ROOT) ;\\\\
	make package ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mbuild\\\\\\\\033[0m. ;\\\\

.PHONY: compile
compile:                      ## Only compile
	@>&2 echo 🤖 Running task \\\\\\\\033[32mcompile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mcompile\\\\\\\\033[0m. ;\\\\

.PHONY: default
default:                      ## Synthesize project files
	@>&2 echo 🤖 Running task \\\\\\\\033[32mdefault\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mdefault\\\\\\\\033[0m. ;\\\\

.PHONY: hello
hello:                        ## No description
	@>&2 echo 🤖 Running task \\\\\\\\033[32mhello\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	echo hello ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mhello\\\\\\\\033[0m. ;\\\\

.PHONY: package
package:                      ## Creates the distribution package
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpackage\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpackage\\\\\\\\033[0m. ;\\\\

.PHONY: post-compile
post-compile:                 ## Runs after successful compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpost-compile\\\\\\\\033[0m. ;\\\\

.PHONY: pre-compile
pre-compile:                  ## Prepare the project for compilation
	@>&2 echo 🤖 Running task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mpre-compile\\\\\\\\033[0m. ;\\\\

.PHONY: test
test:                         ## Run tests
	@>&2 echo 🤖 Running task \\\\\\\\033[32mtest\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mtest\\\\\\\\033[0m. ;\\\\

.PHONY: world
world:                        ## No description
	@>&2 echo 🤖 Running task \\\\\\\\033[32mworld\\\\\\\\033[0m... ;\\\\
	set -e ;\\\\
	cd $(ROOT) ;\\\\
	echo \\"running hello\\" ;\\\\
	cd $(ROOT) ;\\\\
	make hello ;\\\\
	>&2 echo 🤖 Finished task \\\\\\\\033[32mworld\\\\\\\\033[0m. ;\\\\

.PHONY: help
help:                         ## Show help messages for make targets
	@echo \\"\\\\033[1;39mCOMMANDS:\\\\033[0m\\"; grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = \\":.*?## \\"}; {printf \\"\\\\t\\\\033[32m%-30s\\\\033[0m %s\\\\n\\", $$1, $$2}'
"
`;
